cpp/tests/stream_compaction/distinct_count_tests.cpp

```diff
@@ -0,0 +1,368 @@
+}
+
+TYPED_TEST(DistinctCountCommon, TableNoNull)
+{
+  using T = TypeParam;
+
+  auto const input1 = cudf::test::make_type_param_vector<T>(
+    {1, 3, 3, 3, 4, 31, 1, 8, 2, 0, 4, 1, 4, 10, 40, 31, 42, 0, 42, 8, 5, 4});
+  auto const input2 = cudf::test::make_type_param_vector<T>(
+    {3, 3, 3, 4, 31, 1, 8, 5, 0, 4, 1, 4, 10, 40, 31, 42, 0, 42, 8, 5, 4, 1});
+
+  std::vector<std::pair<T, T>> pair_input;
+  std::transform(
+    input1.begin(), input1.end(), input2.begin(), std::back_inserter(pair_input), [](T a, T b) {
+      return std::make_pair(a, b);
+    });
+
+  cudf::test::fixed_width_column_wrapper<T> input_col1(input1.begin(), input1.end());
+  cudf::test::fixed_width_column_wrapper<T> input_col2(input2.begin(), input2.end());
+
+  std::vector<cudf::column_view> cols{input_col1, input_col2};
+  cudf::table_view input_table(cols);
+
+  cudf::size_type const expected =
+    std::set<std::pair<T, T>>(pair_input.begin(), pair_input.end()).size();
+  EXPECT_EQ(expected, cudf::unordered_distinct_count(input_table, null_equality::EQUAL));
+
+  auto const new_end      = std::unique(pair_input.begin(), pair_input.end());
+  auto const gold_ordered = new_end - pair_input.begin();
+  EXPECT_EQ(gold_ordered, cudf::distinct_count(input_table, null_equality::EQUAL));
+}
+
+struct DistinctCount : public cudf::test::BaseFixture {
+};
+
+TEST_F(DistinctCount, WithNull)
+{
+  using T = int32_t;
+
+  std::vector<T> input               = {1,   3,  3,  XXX, 31, 1, 8,  2, 0, XXX, XXX,
+                          XXX, 10, 40, 31,  42, 0, 42, 8, 5, XXX};
+  std::vector<cudf::size_type> valid = {1, 1, 1, 0, 1, 1, 1, 1, 1, 0, 0,
+                                        0, 1, 1, 1, 1, 1, 1, 1, 1, 0};
+
+  cudf::test::fixed_width_column_wrapper<T> input_col(input.begin(), input.end(), valid.begin());
+
+  // explicit instantiation to one particular type (`double`) to reduce build time
+  cudf::size_type const expected = std::set<double>(input.begin(), input.end()).size();
+  EXPECT_EQ(
+    expected,
+    cudf::unordered_distinct_count(input_col, null_policy::INCLUDE, nan_policy::NAN_IS_VALID));
+
+  auto const new_end      = std::unique(input.begin(), input.end());
+  auto const gold_ordered = new_end - input.begin() - 3;
+  EXPECT_EQ(gold_ordered,
+            cudf::distinct_count(input_col, null_policy::EXCLUDE, nan_policy::NAN_IS_VALID));
+}
+
+TEST_F(DistinctCount, IgnoringNull)
+{
+  using T = int32_t;
+
+  std::vector<T> input               = {1,   YYY, YYY, XXX, 31, 1, 8,  2, 0, XXX, 1,
+                          XXX, 10,  40,  31,  42, 0, 42, 8, 5, XXX};
+  std::vector<cudf::size_type> valid = {1, 0, 0, 0, 1, 1, 1, 1, 1, 0, 1,
+                                        0, 1, 1, 1, 1, 1, 1, 1, 1, 0};
+
+  cudf::test::fixed_width_column_wrapper<T> input_col(input.begin(), input.end(), valid.begin());
+
+  cudf::size_type const expected = std::set<T>(input.begin(), input.end()).size();
```

Comment:
This casts from an unsigned `size_t` to a signed `cudf::size_type`. I think that raises a warning but I haven't compiled it to check. Typically I see this written instead:
```suggestion
  auto const expected = static_cast<cudf::size_type>(std::set<T>(input.begin(), input.end()).size());
```

Reply:
I noticed this implicit conversion as well but didn't fix it since I didn't see any warnings. Corrected.